service/league_service/league_rater.py

```diff
@@ -18,4 +18,68 @@ def rate(
         outcome: GameOutcome,
         player_rating: Rating,
     ):
+        boost = 0
+        reduction = 0
+        rating = player_rating[0] - 3 * player_rating[1]
+        player_div = cls._get_player_division(current_score.division_id, league)
+        # Highest divison has lowest id
+        lowest_id, highest_id = cls._get_division_id_range()
+
+        if current_score.game_count >= 10:
+            if current_score.division_id is None:
+                return cls._do_placement(league, current_score, rating)
+
+            if rating > player_div.max_rating:
+                boost = 1
+            elif rating < player_div.min_rating:
+                reduction = 1
+            # Linear interpolation of score and rating to have players in top division sorted by rating
+            elif player_div.id == lowest_id and current_score.score < (player_div.highest_score - player_div.lowest_score) \
+                    * (rating - player_div.min_rating) / (player_div.max_rating - player_div.min_rating):
+                boost = 1
+
+            if outcome == "VICTORY":
+                current_score.score += 1 + boost
+            elif outcome == "DEFEAT":
+                current_score.score -= 1 + reduction
+            if current_score.score < 0:
+                current_score.score = 0
+
+            if current_score.score > player_div.highest_score:
+                if lowest_id < current_score.division_id:
+                    current_score.division_id -= 1
+            elif current_score.score < player_div.lowest_score:
+                if highest_id > current_score.division_id:
+                    current_score.division_id += 1
+
+        current_score.game_count += 1
+
         return current_score
+
+    @classmethod
+    def _get_player_division(cls, division_id, league):
+        for div in league.divisions:
+            if div.id == division_id:
+                return div
+
+    @classmethod
+    def _get_division_id_range(cls, league):
+        highest_id = league.divisions[0].id
+        lowest_id = highest_id
+        for div in league.divisions:
+            if div.id < lowest_id:
+                lowest_id = div.id
+            if div.id > highest_id:
+                highest_id = div.id
+        return lowest_id, highest_id
+
+    @classmethod
+    def _do_placement(cls, league, current_score, rating):
+        rating -= 300
```

Comment:
This `300` could be a parameter in `service.config`.
(https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad/47902#47902)

Reply:
Done